### PR TITLE
feat: add tests in rust

### PR DESCRIPTION
**Pull request by swarnimarun**
use command `cargo t -p rust-tests` to run the end to end tests

TODO:
- [x] PoC for tests written in rust with similar simplicity to pytests
- [x] Port atleast 1 test
- [x] Port rest of the tests from pytest
- [x] Ensure all tests are working
- [ ] Use new rust tests in CI

### CODE REVIEW

**Review by wolfv — APPROVED** [review]

**Comment by wolfv** [comment]
@swarnimarun is this ready for review & merge?

File: rust-tests/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+                    let p = p.as_object().unwrap();
+                    assert!(c["_path"] == p["_path"]);
+                    assert!(c["path_type"] == p["path_type"]);
+                    if p["_path"]
+                        .as_str()
+                        .map(|s| s.contains("dist-info"))
+                        .unwrap_or_default()
+                    {
+                        // TODO: figure out why this is not the same b/w expected and cmp
+                        // assert!(c["sha256"] == p["sha256"]);
```
Comment (swarnimarun) [inline rust-tests/src/lib.rs:309]: everything else is working except, can't figure out why the sha256/hash is different here. hmmm...

**Comment by swarnimarun** [comment]
There is a bug with temp_dir paths, so it's best to run `cargo t -p rust-tests -- --test-threads 1`, note single test thread, means sequential tests rather than parallel tests, pytest matches this behaviour by default afaik, not sure why this sometimes bugs some of the tests but it's a gotcha.

**Comment by wolfv** [comment]
Good to know. Can we set this in the Cargo.toml file as default?

File: rust-tests/src/lib.rs (hunk 1 above)
Comment (wolfv) [reply rust-tests/src/lib.rs:309]: I can have a look!

Comment (wolfv) [reply rust-tests/src/lib.rs]: I think the logic was "backwards" - the distinfo files are the ones that embed the build path and are therefore not deterministic.

**Comment by wolfv** [comment]
I think the tests are not yet executed in CI, correct?

**Pull request merged**